arabic/unpacked/arabic.js

```diff
@@ -0,0 +1,448 @@
+          var texCommand = dict[key][0];
+          definitions[texCommand] = key;
+        });
+
+        return definitions;
+      }()
+    });
+
+
+    TEX.Parse.Augment(function () {
+      var parsers = {};
+
+      Object.keys(dict).forEach(function (key) {
+        parsers[key] = dict[key][1]; // Parser function
+      });
+
+      return parsers;
+    }());
+
+
+    TEX.Parse.Augment({
+      flipHorizontal: function (token) {
+        token.arabicFlipH = !token.arabicFlipH;
+          // Invert the value, because flipping twice means, it is not flipped
+        return token;
+      },
+      arabicNumber: function (token) {
+        var numbersMap = MathJax.Hub.config.Arabic.numbersMap;
+        var text = token.data[0].data[0];
+        var mapped = text.replace(englishNumbersRegExp, function (m) {
+          return numbersMap[m];
+        });
+
+        if (mapped !== text) {
+          token.data[0].data[0] = mapped;
+          token.arabicFontLang = 'ar';
+        }
+
+        return this.flipHorizontal(token);
+      },
+      arabicIdentifier: function (token) {
+        var text = token.data[0].data[0];
+
+        if ('chars' === token.data[0].type) {
+          // English Symbols like X and Y
+          var mapped = text.replace(identifiersKeysRegExp, function (m) {
+            return identifiersMap[m.toLowerCase()];
+          });
+
+          if (mapped !== text) {
+            token.data[0].data[0] = mapped;
+            token.arabicFontLang = 'ar';
+          }
+        }
+
+        return this.flipHorizontal(token);
+      },
+      arabicOperator: function (token) {
+        var text = token.data[0].data[0];
+        var mapped = text.replace(operatorsKeysRegExp, function (m) {
+          return operatorsMap[m];
+        });
+
+        if (mapped !== text) {
+          token = this.flipHorizontal(token);
+          token.arabicFontLang = 'ar';
+          token.data[0].data[0] = mapped;
+        }
+
+        return token;
+      },
+      _getArgumentMML: function (name) {
+        //  returns an argument that is a single MathML element
+        //  (in an mrow if necessary)
+        //
+        //  This functions has been copied here from extensions/TeX/HTML.js, to avoid
+        //  adding a dependency.
+        //
+        //  TODO: Consider importing (as a dependency) this from HTML.js instead!
+        var arg = this.ParseArg(name);
+        if (arg.inferred && arg.data.length === 1) {
+          arg = arg.data[0];
+        } else {
+          delete arg.inferred;
+        }
+
+        return arg;
+      },
+      Push: function () {
+        var beforeLang = this.stack.env.lang;
+        var retVal = texParsePush.apply(this, arguments);
+
+        // Huge giant hack to propagate `lang` from Arrays (and other environments?)
+        // to their children fractions and others.
+        if (beforeLang) {
+          if (!this.stack.env.lang) {
+            this.stack.env.lang = beforeLang;
+          }
+        }
```

Comment:
You are right to be concerned about making a change to this function, since it is used for every element that is included in every equation on the page.  It turns out that the issue with the array that you are trying to resolve is actually a bug in the TeX input jax (I will work on fixing it, but haven't yet).  In the meantime, I'd recommend an alternative that is less intrusive.

The problem is actually occurring in the part of the TeX input jax that clears local changes at the end of a table cell.  This is in the `TeX.Stack.Item.array` object's `clearEnv()` method.  So you could replace that instead of replacing `Push()`, as follows.  Add

```
var array = TEX.Stack.Item.array;
var arrayClearEnv = array.prototype.clearEnv;
array.Augment({
  clearEnv: function () {
    var lang = this.env.lang;
    arrayClearEnv.apply(this,arguments);
    if (lang) this.env.lang = lang;
  }
});
```

after your current line 442, and that should resolve the array issue without having to change the `Push()` method.  So you can then remove lines 374 to 387.